src/rpc/decode.ts

```diff
@@ -0,0 +1,46 @@
+const abiDecoder = require('abi-decoder');
+import ethereum from '../network/providers/ethereum';
+
+type Param = {
+    name: string;
+    value: string;
+    type: string;
+};
+
+type DecodedTxInput = {
+    name: string;
+    params: Param[];
+};
+
+// A map to keep track of which ABIs have been added, to avoid duplicates.
+const addedABIs: Record<string, boolean> = {};
+
+/**
+ * Decode Tx and return human-readable input data.
+ */
+export async function decodeTxInputData(
+    target: string | undefined,
+    contract: string | undefined,
+    params: unknown[]
+): Promise<DecodedTxInput | null> {
+    const txInputData =
+        params && params.length > 0 && (params[0] as Record<string, string>).hasOwnProperty('data')
+            ? (params[0] as Record<string, string>).data
+            : null;
+
+    if (!target || !contract || !txInputData) {
+        return null;
+    }
+
+    // TODO: keep a cache to avoid fethcing same ABI multiple times.
+    const identity = new URL(target).host.replace(/.point$/, '');
+    const {_jsonInterface} = await ethereum.loadWebsiteContract(identity, contract);
+
+    if (!addedABIs[`${target}:${contract}`]) {
+        abiDecoder.addABI(_jsonInterface);
+        addedABIs[`${target}:${contract}`] = true;
+    }
+
+    const decoded = abiDecoder.decodeMethod(txInputData);
+    return decoded ? (decoded as DecodedTxInput) : null;
```

Comment:
What does faulty `decoded` variable mean? Should we consider throwing an error in this case. Or is it a normal scenario?

Reply:
When `abiDecoder.decodeMethod` fails, it returns `undefined`, instead of for example throwing an error.

We could throw an error if we wanted. It comes down to how critical we think it is to not be able to decode the tx input.

My thinking was that it wasn't critical, if for any reason, we couldn't decode it. The confirmation window will show the raw data and users can decide if they want to confirm or not. But if we want to treat it as something critical that should stop the flow, we could do it.

There's the case of interacting with contracts that are not deployed to Point Network. In those cases, we won't be able to fetch the ABI (at least with these changes); so, decoding will not work. I'm trying to think if this is a valid use case though, I'm not sure if it's something plausible.

Reply:
Thanks @germanviescas, of course, I agree it makes sense to let user decide whether they want to proceed with the request or not.

I just thought that internally we still want to know exactly what happens. The `decodeMethod` call may fail due to multiple reasons, either the ABI is outdated or the provided parameters are incorrect or maybe some other reason.

We can integrate the feature early and leave it as is for now, but I expect some issues here, and want to make sure we will be able to debug it.

At least let's log everything we know about the failure. Like, the contract and the method names, the provided args, etc etc.

While we'll be able to debug it, we can still let user make their own choice regarding the tx.

WDYT?